engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -48,70 +48,157 @@ object WorkflowStore {
-
-  override def receive = {
-    case SubmitWorkflow(source) =>
-      val id = add(NonEmptyList(source)).head
-      registerIdWithMetadataService(id)
-      sender ! WorkflowSubmittedToStore(id)
-      log.info(s"Workflow $id submitted.")
-    case BatchSubmitWorkflows(sources) =>
-      val ids = add(sources)
-      ids foreach registerIdWithMetadataService
-      sender ! WorkflowsBatchSubmittedToStore(ids)
-      log.info(s"Workflows ${ids.list.mkString(", ")} submitted.")
-    case FetchRunnableWorkflows(n) => sender ! newWorkflowMessage(n)
-    case RemoveWorkflow(id) =>
-      if (!remove(id)) log.info(s"Attempted to remove ID $id from the WorkflowStore but it already exists!")
+case class WorkflowStoreActor(store: WorkflowStoreDatabase) extends LoggingFSM[WorkflowStoreActorState, WorkflowStoreActorData] with ActorLogging with ServiceRegistryClient {
+
+  implicit val ec: ExecutionContext = context.dispatcher
+
+  startWith(Idle, WorkflowStoreActorData(None, List.empty))
+  self ! InitializerCommand
+
+  when(Idle) {
+    case Event(cmd @ SubmitWorkflow(source), _) =>
+      val sndr = sender
+      val work = store.add(NonEmptyList(source)) map { ids =>
+        val id = ids.head
+        registerIdWithMetadataService(id)
+        sndr ! WorkflowSubmittedToStore(id)
+        log.info("Workflow {} submitted.", id)
+      }
+      addWorkCompletionHooks(cmd, work)
+      goto(Working) using stateData.copy(currentOperation = Option(cmd))
+    case Event(cmd @ BatchSubmitWorkflows(sources), _) =>
+      val sndr = sender
+      val work = store.add(sources) map { ids =>
+        val id = ids.head
+        ids foreach registerIdWithMetadataService
+        sndr ! WorkflowsBatchSubmittedToStore(ids)
+        log.info("Workflows {} submitted.", ids.list.mkString(", "))
+      }
+      addWorkCompletionHooks(cmd, work)
+      goto(Working) using stateData.copy(currentOperation = Option(cmd))
+    case Event(cmd @ FetchRunnableWorkflows(n), _) =>
+      val sndr = sender
+      val work = newWorkflowMessage(n) map { nwm =>
+        nwm match {
+          case NewWorkflowsToStart(workflows) => log.info("{} workflows fetched", workflows.size)
+          case NoNewWorkflowsToStart => log.info("No workflows fetched")
+          case _ => log.error("Unexpected response from newWorkflowMessage({}): {}", n, nwm)
+        }
+        sndr ! nwm
+      }
+      addWorkCompletionHooks(cmd, work)
+      goto(Working) using stateData.copy(currentOperation = Option(cmd))
+    case Event(cmd @ RemoveWorkflow(id), _) =>
+      val work = store.remove(id) map { removed =>
+        if (removed) {
+          log.info("Workflow {} removed from store successfully.", id)
+        } else {
+          log.warning(s"Attempted to remove ID {} from the WorkflowStore but it didn't exist", id)
+        }
+      }
+      addWorkCompletionHooks(cmd, work)
+      goto(Working) using stateData.copy(currentOperation = Option(cmd))
+    case Event(InitializerCommand, _) =>
```

Comment:
Hmm it just feels like the FSM is in a distinct uninitialized state in which it can't actually be used until this work completes.